[jit] register __getitem__ builtin

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22276 [jit] register __getitem__ builtin**

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

Differential Revision: [D16060595](https://our.internmc.facebook.com/intern/diff/D16060595)

wanchaol committed:
[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

zdevito reviewed (commented):
Cool -- I think this change is a good thing. There are a few places where the code is duplicated now that everything goes through __getitem__

wanchaol committed:
Update on "[jit] register __getitem__ builtin"

[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

gh-metadata: pytorch pytorch 22276 gh/wanchaol/28/head

wanchaol committed:
rebase and address comments on "[jit] register __getitem__ builtin"

[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

gh-metadata: pytorch pytorch 22276 gh/wanchaol/28/head

wanchaol commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -2880,7 +2880,13 @@ struct to_ir {
       auto* idx = emitExpr(subscript_exprs[0]);
       return emitBuiltinCall(
-          loc, *graph, aten::select, c10::nullopt, {gatherable, idx}, {}, true);
+          loc,
```
Merged them together, I also tried to merge the tuple indexing, but it might resides on multiple sugared values so I decided to still leave it alone.

wanchaol committed:
Update on "[jit] register __getitem__ builtin"

[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

gh-metadata: pytorch pytorch 22276 gh/wanchaol/28/head

wanchaol committed:
Update on "[jit] register __getitem__ builtin"

[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

gh-metadata: pytorch pytorch 22276 gh/wanchaol/28/head

zdevito reviewed (approved):
Looks good now

wanchaol committed:
Update on "[jit] register __getitem__ builtin"

[jit] register __getitem__ builtin

Summary:

Follow up of https://github.com/pytorch/pytorch/pull/21990, I am switching index select operations to a standard __getitem__ builtin, rather than bunch of different builtins according to the type, such as prim::DictIndex, prim::ListIndex, etc. This will also aligned with the some other magic methods that we already use

gh-metadata: pytorch pytorch 22276 gh/wanchaol/28/head
